lib/ts/recipe/passwordless/api/implementation.ts

```diff
@@ -37,112 +41,208 @@ export default function getAPIImplementation(): APIInterface {
-                });
-            }
+                let isFirstFactor = session === undefined;
+                if (shouldCreateSession) {
+                    session = await Session.createNewSession(
+                        input.options.req,
+                        input.options.res,
+                        input.tenantId,
+                        loginMethod.recipeUserId,
+                        {},
+                        {},
+                        input.userContext
+                    );
+                }
 
-            const session = await Session.createNewSession(
-                input.options.req,
-                input.options.res,
-                input.tenantId,
-                loginMethod.recipeUserId,
-                {},
-                {},
-                input.userContext
-            );
+                if (session === undefined) {
+                    throw new Error("should never come here");
+                }
 
-            return {
-                status: "OK",
-                createdNewRecipeUser: response.createdNewRecipeUser,
-                user: response.user,
-                session,
-            };
+                if (mfaInstance !== undefined) {
+                    let sessionRes = await mfaInstance.updateSessionAndUserAfterFactorCompletion({
+                        session,
+                        isFirstFactor,
+                        factorId: factorId,
+                        userInfoOfUserThatCompletedSignInOrUpToCompleteCurrentFactor: {
+                            createdNewUser: response.createdNewRecipeUser,
+                            recipeUserId: response.recipeUserId,
+                            user: response.user,
+                        },
+                        userContext: input.userContext,
+                    });
+
+                    if (sessionRes.status === "MFA_FLOW_ERROR") {
+                        return {
+                            status: "SIGN_IN_UP_FAILED",
+                            reason: sessionRes.reason,
+                        };
+                    }
+
+                    if (sessionRes.status === "RECURSE_FOR_RACE_CONDITION") {
+                        continue;
+                    }
+
+                    let user = await getUser(response.user.id, input.userContext);
+
+                    if (user === undefined) {
+                        throw new SessionError({
+                            type: SessionError.UNAUTHORISED,
+                            message: "Session user not found",
+                        });
+                    }
+
+                    response.user = user;
+                }
+
+                return {
+                    status: "OK",
+                    session,
+                    createdNewRecipeUser: response.createdNewRecipeUser,
+                    user: response.user,
+                };
+            }
         },
+
         createCodePOST: async function (input) {
```

Comment:
shouldnt we check for validateForMultifactorAuthBeforeFactorCompletion in this api too? So that users see the relevant error early on as opposed to seeing it when they enter the otp?

Reply:
Yes, done. Added the check to totp create device as well

Reply:
we can't do this in create code as we do not know the factorId at this point. It will be handled in consume code

Reply:
Now that we will be adding factor id as an input, will we do this check in this API? What had we decided?

Reply:
We need to check them cause we don't want to send unnecessary emails / text messages if the factor setup is not allowed etc... These will be different support codes compared to the ones in consumeCode, even though semantically they mean the same case

Reply:
@porcellus